tests/lib/openstack_query/test_query_parser.py

```diff
@@ -0,0 +1,415 @@
+        Tests that run_sort functions expectedly - with one sorting key
+        Should call run_sort method which should get the appropriate sorting
+        key lambda function and sort dict accordingly
+        """
+
+        mock_obj_list = [
+            {"arg1": "a", "arg2": 1},
+            {"arg1": "b", "arg2": 1},
+            {"arg1": "a", "arg2": 2},
+            {"arg1": "b", "arg2": 2},
+        ]
+
+        res = self.instance._run_sort(mock_obj_list, sort_by_specs=mock_sort_by_specs)
+        self.assertEqual(res, expected_res)
+
+    @parameterized.expand(
+        [
+            (
+                "boolean ascending",
+                {"enabled": False},
+                [
+                    {"enabled": False},
+                    {"enabled": True},
+                ],
+            ),
+            (
+                "boolean ascending",
+                {"enabled": True},
+                [
+                    {"enabled": True},
+                    {"enabled": False},
+                ],
+            ),
+        ]
+    )
+    def test_run_sort_with_boolean(self, _, mock_sort_by_specs, expected_res):
+        mock_obj_list = [
+            {"enabled": False},
+            {"enabled": True},
+        ]
+
+        res = self.instance._run_sort(mock_obj_list, sort_by_specs=mock_sort_by_specs)
+        self.assertEqual(res, expected_res)
+
+    @parameterized.expand(
+        [
+            ("group by arg1", "arg1", ["a", "b", "c"]),
+            ("group by arg2", "arg2", [1, 2, 3]),
+        ]
+    )
+    def test_build_unique_val_groups(self, _, mock_group_by_prop, expected_unique_vals):
+        """
+        Tests that build_unique_val_groups method functions expectedly
+        method should find all unique values for a test object list for a given property and create appropriate
+        group_ranges to be used for group-by
+        """
+
+        def mock_get_prop(obj, prop):
+            return obj[prop.name]
+
+        self.mock_prop_handler.get_prop = MagicMock(wraps=mock_get_prop)
+
+        obj_list = [
+            {"arg1": "a", "arg2": 1},
+            {"arg1": "b", "arg2": 2},
+            {"arg1": "c", "arg2": 3},
+        ]
+        mock_prop_enum = MagicMock()
+        mock_prop_enum.name = mock_group_by_prop
+        res = self.instance._build_unique_val_groups(
+            obj_list,
+            mock_prop_enum,
+        )
+        expected_group_names = [
+            f"{mock_group_by_prop} with value {val}" for val in expected_unique_vals
+        ]
+        self.assertEqual(list(res.keys()), expected_group_names)
+
+        # check that groups are unique
+        for i, key in enumerate(res.keys()):
+            for j, test_val in enumerate(expected_unique_vals):
+                test_bool = res[key]({mock_group_by_prop: test_val})
+
+                # if the function is run with prop matching expected val - should return True
+                if i == j:
+                    self.assertTrue(test_bool)
```

Comment:
This is a really complex way to test this, I would simply do:

```
assert len(set(res) == len(expected_group_names)
assert all(i for i in rest in expected_group_names)
```

Also since all the values don't overlap (e.g. a,b,c and 1,2,3) this wouldn't remove anything AFAIK?

Reply:
seems dubious - you're relying on `i` being in the group name but not actually checking if the function that it maps to works

Reply:
My concern is us doing two steps in our tests:

```
        res = self.instance._build_unique_val_groups(
            obj_list,
            mock_prop_enum,
        )
```

Then instead of directly checking `res` we appear to mutate it, I guess it's this syntax I'm really struggling with:

`test_bool = res[key]({mock_group_by_prop: test_val})`

I can see we iterate the key, so we have something like a group of things

But then it's this bit where we effectively do

`res[key]({"arg1: ["a", "b", "c"]})` that's causing my problems

IMO it's too clever for it's own good, I've took a few attempts to parse what's going on (for example what are we calling with the `(DICT)`, since a dictionary should return a data structure not another callable?

Whilst I understand it's an internal data structure I worry we're writing code that's so dense in meaning only a few people will be able to maintain it?

Reply:
I've simplified this test - now we just run the group mappings and check the output matches expected